### PR TITLE
[Documentation] Cleanup Docs and Workflows

### PR DESCRIPTION
**Pull request by h5law**
### Human Summary

This PR moves **all** documentation to the `docs/` subdir and updates the workflow files. It also makes codecov informational only instead of passing/failing CI checks.

### Summary generated by Reviewpad on 03 Jan 24 14:06 UTC
This pull request includes the following changes:

1. Ignoring specific files and directories: The ".DS_Store" file, which are Mac OS X system files, and the ".idea/" directory, which contains Goland and JetBrains IDE files, are now being ignored. Additionally, a new file called ".tool-versions" is also being ignored, which is a local tool-versions file for asdf.

2. Changes to the `.github/codecov.yml` file: This file has been added with a mode of 100644. The `coverage` section has been added with the `status` configuration. The `informational` property has been set to `true` for both the `project` and `patch` sections.

3. Renaming and content changes for the `SMT.md` file: The file has been renamed to `smt.md` and includes an added section on "Closest Proof Use Cases." The "Compression" and "Data Loss" sections have been updated. An additional change is the renaming of documentation links for "KVStore" and "MerkleSumTrie" to lowercase.

4. Changes to the `docs/KVStore.md` file: This file has undergone a 90% similarity index change and has been renamed to `kvstore.md`. The changes involve reformatting the indentation of subheadings under the "Implementation" and "Accessor Methods" sections.

5. Changes to the `MerkleSumTrie.md` file: The filename has been changed to `merkle-sum-trie.md`. There have been changes in the headings and subheadings, including indentation adjustments. A file link for the `SMT documentation` has also been updated.

6. Updates to the `test.yml` file: The workflow name has been changed, the value of an environment variable has been updated, and various modifications have been made to different steps.

Please review these changes and ensure that they align with the intended modifications.

## Issue

N/A

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [x] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make test_all`
- [ ] **Run all/relevant benchmarks (if optimising)**: `make benchmark_{all | suite name}`

## Required Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code ([`godoc` format comments](https://go.dev/blog/godoc) see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))

### If Applicable Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated any relevant README(s)/documentation and left TODOs throughout the codebase
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

### CODE REVIEW

File: .github/codecov.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Only post coverage comments and not pass/fail reports in the CI checks
```
Comment (Olshansk) [inline .github/codecov.yml:1]: Do you think we need this at all given that we're not really acting on the feedback?

File: docs/smt.md [hunk 2]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
-can only happen due to the method's deterministic behaviour and the prover
-knowing the hash before they have "closed" the trie. The intended use of this
-method is that the verifier gives the hash only after the prover has closed their
-trie and submitted the closed trie's root hash. This enables the verifier to
-verify the integrity of the proof (if the trie was changed the root hash would
-be different) and also guarantees the pseudo-random proof of inclusion was not
-a maliciously placed leaf.
+the use case of this function this may expose a vulnerability. **It is not
+intendend to be used as a general purpose proof mechanism**, but instead as a
+**Commit and Reveal** mechanism, as detailed below.
```
Comment (Olshansk) [inline docs/smt.md:390]: 🔥

File: docs/smt.md [hunk 3]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
+referred to as closing the trie, such that no more updates are made to it once
+committed._
+
+Take the following attack vector (**without** a commit prior to a reveal) into
+consideration:
+
+1. The **prover** identifies the hash the **verifier** intends to check against
+1. The **prover** then places a leaf in the tree with a common prefix to the
+   identified hash
+1. Due to the deterministic nature of the `ClosestProof` method this leaf will
```
Comment (Olshansk) [inline docs/smt.md:407]:
```suggestion
1. Due to the deterministic nature of the `ClosestProof`, method this leaf will
```

File: docs/smt.md [hunk 4]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
+_NOTE: Throughout this document, `commitment` of the the trie's root hash is also
+referred to as closing the trie, such that no more updates are made to it once
+committed._
+
+Take the following attack vector (**without** a commit prior to a reveal) into
+consideration:
+
+1. The **prover** identifies the hash the **verifier** intends to check against
+1. The **prover** then places a leaf in the tree with a common prefix to the
+   identified hash
```
Comment (Olshansk) [inline docs/smt.md:406]:
```suggestion
1. The **prover** inserts a leaf (i.e. a value) whose key (determined via the hasher) has a
   longer common prefix than any other leaf in the trie.
```

File: docs/smt.md [hunk 5]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
+Where there are two actors involved, the **prover** and **verifier**.
+
+_NOTE: Throughout this document, `commitment` of the the trie's root hash is also
+referred to as closing the trie, such that no more updates are made to it once
+committed._
+
+Take the following attack vector (**without** a commit prior to a reveal) into
+consideration:
+
+1. The **prover** identifies the hash the **verifier** intends to check against
```
Comment (Olshansk) [inline docs/smt.md:404]:
```suggestion
1. The **verifier** picks the hash (i.e. a single branch) they intend to check
1. The **prover** identifies the hash the **verifier** intends to check against
```

File: docs/smt.md [hunk 6]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
+1. The **prover** then places a leaf in the tree with a common prefix to the
+   identified hash
+1. Due to the deterministic nature of the `ClosestProof` method this leaf will
+   **always** be returned given the identified hash.
+1. The **verifier** then verifies the revealed `ClosestProof`, in turn validating
+   a maliciously placed leaf.
```
Comment (Olshansk) [inline docs/smt.md:410]:
```suggestion
1. The **verifier** verifies the revealed `ClosestProof`, which returns a branch
    the **prover** inserted after knowing which leaf was going to be checked.
```

File: docs/smt.md [hunk 7]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
+#### Closest Proof Use Cases
+
+The `CloestProof` function is intended for use as a `commit & reveal` mechanism.
+Where there are two actors involved, the **prover** and **verifier**.
+
+_NOTE: Throughout this document, `commitment` of the the trie's root hash is also
+referred to as closing the trie, such that no more updates are made to it once
+committed._
+
+Take the following attack vector (**without** a commit prior to a reveal) into
```
Comment (Olshansk) [inline docs/smt.md:401]:
```suggestion
Consider the following attack vector (**without** a commit prior to a reveal) into
```

File: docs/smt.md [hunk 8]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
+
+1. The **prover** identifies the hash the **verifier** intends to check against
+1. The **prover** then places a leaf in the tree with a common prefix to the
+   identified hash
+1. Due to the deterministic nature of the `ClosestProof` method this leaf will
+   **always** be returned given the identified hash.
+1. The **verifier** then verifies the revealed `ClosestProof`, in turn validating
+   a maliciously placed leaf.
+
+Take the following normal flow (**with** a commit prior to reveal) as
```
Comment (Olshansk) [inline docs/smt.md:412]:
```suggestion
Consider the following normal flow (**with** a commit prior to reveal) as
```

File: docs/smt.md [hunk 9]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
+
+1. The **prover** commits to the state of their trie and publishes their root
```
Comment (Olshansk) [inline docs/smt.md:414]:
```suggestion
1. The **prover** commits to the state of their trie by publishing their root
```

File: docs/smt.md [hunk 10]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
+   hash, _closing_ their trie.
```
Comment (Olshansk) [inline docs/smt.md:415]:
```suggestion
   hash, thereby _closing_ their trie and not being able to make further changes.
```

File: docs/smt.md [hunk 11]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
+1. The **verifier** provides a hash to be used in the `commit & reveal` process
```
Comment (Olshansk) [inline docs/smt.md:416]:
```suggestion
1. The **verifier** selects a hash to be used in the `commit & reveal` process
    that the **prover** must provide a closest proof for.
```

File: docs/smt.md [hunk 12]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
+1. The **prover** then utilises this hash and the `ClosestProof` method on their
+   _closed_ trie, producing a `ClosestProof`, thus revealing a deterministic,
+   yet pseudo-random leaf.
```
Comment (Olshansk) [inline docs/smt.md:419]:
```suggestion
1. The **prover** utilises this hash and computes the `ClosestProof` on their
   _closed_ trie, producing a `ClosestProof`, thus revealing a deterministic,
   yet pseudo-random leaf that existed in the tree prior to commitment.
```

File: docs/smt.md [hunk 13]
```diff
@@ -384,21 +385,40 @@ of the trie.
...
+1. The **verifier** then verifies the proof, in turn verifying the commitment
+   made by the **prover** and the state of the **prover**'s trie.
```
Comment (Olshansk) [inline docs/smt.md:421]:
```suggestion
1. The **verifier** verifies the proof, in turn, verifying the commitment
   made by the **prover** to the state of the trie in the first step.
1. The **prover** had no opportunity to insert a new leaf into the trie 
    after learning which hash the **verifier** was going to require a
    `ClosestProof` for.
```

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: .github/codecov.yml (hunk 1 above)
Comment (h5law) [reply .github/codecov.yml:1]: I like the coverage percentages, shows the lib has good test coverage

Comment (h5law) [reply .github/codecov.yml:1]: I use the feedback when I add new features

Comment (Olshansk) [reply .github/codecov.yml:1]: > I like the coverage percentages, shows the lib has good test coverage

👍 

> I use the feedback when I add new features

When you put together a contribution guide for this repo, make sure this is clear.

As a n00b, I'd probably think `this tool I've never seen or used is giving me a ⚠️. Do I ignore it or keep going?`

**Review by Olshansk — APPROVED** [review]

**Pull request merged**